dali/imgcodec/decoders/tiff_libtiff.cc

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <tiffio.h>
+#include "dali/imgcodec/decoders/tiff_libtiff.h"
+#include "dali/imgcodec/util/convert.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/kernels/common/utils.h"
+
+#define LIBTIFF_CALL_SUCCESS 1
+#define LIBTIFF_CALL(call)                                \
+  do {                                                    \
+    int retcode = (call);                                 \
+    DALI_ENFORCE(LIBTIFF_CALL_SUCCESS == retcode,         \
+      "libtiff call failed with code "                    \
+      + std::to_string(retcode) + ": " #call);            \
+  } while (0)
+
+namespace dali {
+namespace imgcodec {
+
+namespace detail {
+
+class DecoderHelper {
+ public:
+  explicit DecoderHelper(ImageSource *in) : stream_(in->Open()), in_(in) {}
+
+  static tmsize_t read(thandle_t handle, void *buffer, tmsize_t n) {
+    DecoderHelper *helper = reinterpret_cast<DecoderHelper *>(handle);
+    return helper->stream_->Read(buffer, n);
+  }
+
+  static tmsize_t write(thandle_t, void *, tmsize_t) {
+    // Not used for decoding.
+    return 0;
+  }
+
+  static toff_t seek(thandle_t handle, toff_t offset, int whence) {
+    DecoderHelper *helper = reinterpret_cast<DecoderHelper *>(handle);
+    helper->stream_->SeekRead(offset, whence);
+    return helper->stream_->TellRead();
+  }
+
+  static int map(thandle_t handle, void **base, toff_t *size) {
+    // This function will be used by LibTIFF only if input is InputKind::HostMemory.
+    DecoderHelper *helper = reinterpret_cast<DecoderHelper *>(handle);
+    if (helper->in_->Kind() != InputKind::HostMemory)
+      std::logic_error("DecoderHelper::map can only be used for InputKind::HostMemory");
```

Comment:
Doesn't this just construct the value?
Can we throw from those functions?
Maybe we should return non-zero value?

Reply:
> Doesn't this just construct the value?

Of course it does. Sorry!

> Can we throw from those functions?

Another great question and I have no idea

> Maybe we should return non-zero value?

Definetely, thanks!

Reply:
Fixed